target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr1043nd-v4.c

```diff
@@ -0,0 +1,191 @@
+#include "nvram.h"
+
+#define TL_WR1043_V4_GPIO_BTN_RESET		2
+#define TL_WR1043_V4_GPIO_BTN_RFKILL		5
+
+#define TL_WR1043_V4_GPIO_LED_WLAN		19
+#define TL_WR1043_V4_GPIO_LED_USB		7
+#define TL_WR1043_V4_GPIO_LED_WPS		1
+#define TL_WR1043_V4_GPIO_LED_SYSTEM		6
+
+#define TL_WR1043_V4_GPIO_USB_POWER		8
+
+#define TL_WR1043_V4_GPIO_LED_WAN		15
+#define TL_WR1043_V4_GPIO_LED_LAN1		9
+#define TL_WR1043_V4_GPIO_LED_LAN2		14
+#define TL_WR1043_V4_GPIO_LED_LAN3		21
+#define TL_WR1043_V4_GPIO_LED_LAN4		20
+
+#define TL_WR1043_V4_KEYS_POLL_INTERVAL		20 /* msecs */
+#define TL_WR1043_V4_KEYS_DEBOUNCE_INTERVAL	(3 * TL_WR1043_V4_KEYS_POLL_INTERVAL)
+
+#define TL_WR1043_V4_FLASH_END		0x1fffffff
+#define TL_WR1043_V4_MAC_OFFSET		0x7fe8b
+
+#define TL_WR1043_V4_EEPROM_ADDR		0x1fff0000
+#define TL_WR1043_V4_WMAC_CALDATA_OFFSET	0x1000
+
+static struct gpio_led tl_wr1043nd_v4_leds_gpio[] __initdata = {
+	{
+		.name		= "tp-link:green:wps",
+		.gpio		= TL_WR1043_V4_GPIO_LED_WPS,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:green:system",
+		.gpio		= TL_WR1043_V4_GPIO_LED_SYSTEM,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:green:wlan",
+		.gpio		= TL_WR1043_V4_GPIO_LED_WLAN,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:green:usb",
+		.gpio		= TL_WR1043_V4_GPIO_LED_USB,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:green:wan",
+		.gpio		= TL_WR1043_V4_GPIO_LED_WAN,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:green:lan1",
+		.gpio		= TL_WR1043_V4_GPIO_LED_LAN1,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:green:lan2",
+		.gpio		= TL_WR1043_V4_GPIO_LED_LAN2,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:green:lan3",
+		.gpio		= TL_WR1043_V4_GPIO_LED_LAN3,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tp-link:green:lan4",
+		.gpio		= TL_WR1043_V4_GPIO_LED_LAN4,
+		.active_low	= 1,
+	},
+};
+
+static struct gpio_keys_button tl_wr1043nd_v4_gpio_keys[] __initdata = {
+	{
+		.desc		= "Reset button",
+		.type		= EV_KEY,
+		.code		= KEY_RESTART,
+		.debounce_interval = TL_WR1043_V4_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= TL_WR1043_V4_GPIO_BTN_RESET,
+		.active_low	= 1,
+	},
+	{
+		.desc		= "RFKILL button",
+		.type		= EV_KEY,
+		.code		= KEY_RFKILL,
+		.debounce_interval = TL_WR1043_V4_KEYS_DEBOUNCE_INTERVAL,
+		.gpio		= TL_WR1043_V4_GPIO_BTN_RFKILL,
+		.active_low	= 1,
+	},
+};
+
+static struct ar8327_pad_cfg tl_wr1043nd_v4_ar8327_pad0_cfg = {
+	.mode = AR8327_PAD_MAC_SGMII,
+	.sgmii_delay_en = true,
+};
+
+static struct ar8327_platform_data tl_wr1043nd_v4_ar8327_data = {
+	.pad0_cfg = &tl_wr1043nd_v4_ar8327_pad0_cfg,
+	.port0_cfg = {
+		.force_link = 1,
+		.speed = AR8327_PORT_SPEED_1000,
+		.duplex = 1,
+		.txpause = 1,
+		.rxpause = 1,
+	},
+};
+
+static struct mdio_board_info tl_wr1043nd_v4_mdio0_info[] = {
+	{
+		.bus_id = "ag71xx-mdio.0",
+		.phy_addr = 0,
+		.platform_data = &tl_wr1043nd_v4_ar8327_data,
+	},
+};
+
+static void __init tl_wr1043nd_v4_setup(void)
+{
+	u8 *mac = (u8 *) KSEG1ADDR(TL_WR1043_V4_FLASH_END - TL_WR1043_V4_MAC_OFFSET);
```

Comment:
@Brother-Lal, please take a look how it's done ex. with ART partition, on other TP-Link devices like [TL-MR3x20](https://github.com/lede-project/source/blob/master/target/linux/ar71xx/files/arch/mips/ath79/mach-tl-mr3x20.c#L81). In most of them, ART data resides in last 64 KB sector in FLASH, at offset _0x1000_ (in that sector).

As the SPI FLASH memory area in QC/A WiSoCs wraps around, you can just set the offset to `0x1fff1000` (it's exactly _0x20000000 - 0x10000 + 0x1000_ -> _SPI FLASH memory address + 16 MB - 64 KB + 4 KB_) and no matter what FLASH size (up to 16 MB) is used, this offset will always point to the _0x1000_ offset inside last 64 KB sector. TL-MR3220 has 4 MB FLASH chip by default.

IMHO, using negative offset here might be confusing for others. Personally, I would stick to the general convention like in many other ar71xx mach-*.c files.

Cheers!